ATHOS-15 Cria tabela total de horas consolidadas por projeto front integracao

caiqueBTC opened the pull request:
# Relatório de Atividade - Implementação

## Descrição da Implementação

Foi implementada a tela de relatório de Atividade que exibe uma tabela consolidada de horas por desenvolvedor e projeto, além de cards com o resumo total de horas por projeto. A integração entre front-end e back-end foi realizada com HTMX, permitindo atualizações dinâmicas dos dados sem recarregar a página.

## Alterações Realizadas

### View (`apps/relatorios/atividade/views.py`)
- Criada view `index` para renderizar a página principal com filtros de data
- Criada view `relatorio_tabela_e_cards` para lidar com requisições HTMX
- Implementada filtragem de registros de `ControleHorasEquipe` baseada em mês/ano
- Realizada agregação de dados para calcular totais de horas por funcionário e projeto
- Preparado contexto com dados para tabela e cards de resumo com percentuais

### Templates (`apps/relatorios/atividade/templates/atividade/`)
- `index.html`: Página principal com formulário de filtros (atributos hx-get e hx-target)
- `partials/_tabela_e_cards.html`: Template parcial para renderização dinâmica da tabela e cards

### URLs (`apps/relatorios/atividade/urls.py`)
- Configuradas rotas para a aplicação de atividade
- Mapeamento de URLs para as views `index` e `relatorio_tabela_e_cards`

## Como Testar

Acesse: http://127.0.0.1:8000/relatorios/atividade/

### Verificações

1. **Filtros e Carregamento Dinâmico**
   - Ao selecionar novo mês/ano, a tabela e cards atualizam automaticamente
   - Períodos sem dados exibem mensagem apropriada

2. **Validação da Tabela de Horas**
   - Células exibem corretamente horas por desenvolvedor/projeto
   - Coluna "Total" soma horas por desenvolvedor
   - Linha "TOTAL GERAL" soma horas por projeto
   - Canto inferior direito mostra soma total de horas

3. **Validação dos Cards de Resumo**
   - Total de horas corresponde ao valor na tabela
   - Percentual calculado corretamente em relação ao total
   - Número de desenvolvedores corresponde à quantidade real

## Dados para Teste no Banco

```sql
-- Limpa os dados existentes para garantir um teste limpo (OPCIONAL)
DELETE FROM controle_horas_equipe;
DELETE FROM funcionario;
DELETE FROM projeto;
DELETE FROM cargo;

-- Inserir Cargos
INSERT INTO cargo (id, sigla) VALUES
(1, 'DEV'),
(2, 'QA');

-- Inserir Projetos
INSERT INTO projeto (id, nome, data_criacao) VALUES
(1, 'Projeto X', '2025-01-10'),
(2, 'Projeto Alpha', '2025-02-15'),
(3, 'Projeto Beta', '2025-03-20'),
(4, 'Doit', '2025-04-01');

-- Inserir Funcionários
INSERT INTO funcionario (id, nome, time, cargo_id, data_criacao) VALUES
(1, 'ANA SILVA', 'Time A', 1, '2025-01-01'),
(2, 'João Silva', 'Time A', 1, '2025-01-01'),
(3, 'Maria Santos', 'Time B', 1, '2025-02-01'),
(4, 'Ronald', 'Time B', 1, '2025-02-01');

-- Inserir Controle de Horas equipe
INSERT INTO controle_horas_equipe_resumo (id, total_dev, total_projeto) VALUES
(1, 360.50, 240.00),  -- January summary
(2, 341.00, 190.25); -- February summary

-- Inserir Controle de horas equipe
INSERT INTO controle_horas_equipe (id, mes, horas, resumo_id, funcionario_id, projeto_id) VALUES
(1, '2025-01-01', 160.00, 1, 1, 1), -- João worked full-time on Project 1001 in Jan 2025
(2, '2025-01-01', 120.50, 1, 2, 2), -- Maria part-time on Project 1001
(3, '2025-01-01', 80.00, 1, 3, 1),  -- Carlos half-time on Project 1002
(4, '2025-02-01', 150.75, 2, 4, 2), -- João continues on Project 1001 in Feb
(5, '2025-02-01', 100.25, 2, 1, 1), -- Maria shifts to Project 1003
(6, '2025-02-01', 90.00, 2, 2, 2);  -- Ana joins Project 1002
```

caiqueBTC committed:
ATHOS-15 feat(relatorios): Cria tabela de horas consolidadas por projeto

Desenvolve a visualização do relatório de atividades, exibindo uma tabela com o total de horas de cada colaborador distribuído por projeto.

Os dados foram mockados diretamente na view para permitir o desenvolvimento e a integração do frontend sem a necessidade de depender da implementação final do banco de dados.

caiqueBTC committed:
ATHOS-15 style(front): Finaliza ajustes de layout do relatorio

caiqueBTC committed:
Merge branch 'develop' into ATHOS-15/Criar-tabela-total-de-horas-consolidadas-por-projeto-front-integracao

caiqueBTC committed:
ATHOS-15 style(front): Finaliza a estilização e interatividade do relatório

A interface do relatório de horas foi completamente estilizada para corresponder ao design de referência, incluindo filtros, tabela e cartões.

A interatividade do filtro de data (mês/ano) foi implementada utilizando HTMX. A escolha por esta tecnologia foi motivada pela sua capacidade de criar uma experiência de usuário dinâmica, sem recarregamento de página, mantendo a lógica de renderização no back-end com Django. Esta abordagem simplifica o desenvolvimento e a manutenção futura.

caiqueBTC committed:
ATHOS-15 refactor(TABELA DE HORAS): remove codigo redudantes verificação de banco

caiqueBTC committed:
ATHOS-15 fix(TABELA DE HORAS) arrumando url duplicadas

caiqueBTC committed:
ATHOS-15 fix: Resolve conflito de merge em urls.py

caiqueBTC committed:
ATHOS-15 fix: Configurando GET deixando mais seguro

caiqueBTC committed:
ATHOS-15 refactor - removendo e renomeando imagens evitar conflitos

caiqueBTC committed:
fix: Resolve conflito de merge em urls.py

RuthMira reviewed (changes requested):
<img width="1907" height="463" alt="image" src="https://github.com/user-attachments/assets/b33254cb-a8c1-4d3b-a1f2-dc25f2056aac" />

caiqueBTC committed:
ATHOS-15 fix(PR): Resolve conflito de merge e corrige hotspots de segurança do HTMX que fazia desaparecimento da tabela

caiqueBTC commented:
Corrigi configurações do HTMX, adicionando arquivo JavaScript interno 
<img width="1594" height="890" alt="Captura de tela de 2025-09-18 00-17-27" src="https://github.com/user-attachments/assets/ac169cf4-71db-4321-b767-ed291d964249" />

caiqueBTC committed:
ATHOS-15 refactor(urls): Simplifica roteamento e remove URL de API duplicada

RuthMira reviewed (dismissed):
Revisar nome para PR , seguir o padrão definido pela equipe

Front não deve expor ou fazer requisições direto para o banco , temos a camada  de service que realiza as funções e a view para se comunicar com o front (acredito que precise de ajuste)

Verificar a necessidade de particionar tabela em uma pasta separada do index.

A parte de filtro não faz parte da sua tasks, como já criou não tem problema , a logica e regra de negocio sempre fica no service 
apps/relatorios/atividade/templatetags/custom_filters.py

No config/settings.py - organizar o django_htmx para o local correto

Aqui foi retirado o registro do caminho atividade/ config/urls.py , não sei como está funcionando se foi registrado dentro da view ou outro local

caiqueBTC committed:
ATHOS-15 refactor(relatorios): Adiciona url na pasta config  e separa lógica em camada de serviço

caiqueBTC commented:
Alterações Implementadas:
Criação da Camada de Serviço: Conforme a sua principal sugestão, toda a lógica de negócio (consultas, filtros e cálculos) foi movida da view para uma nova camada de serviço em apps/relatorios/atividade/services.py. A view agora tem a responsabilidade única de intermediar a requisição e renderizar o template.

Ajuste na Estrutura de URLs: A estrutura de roteamento foi ajustada. Agora, as URLs de cada sub-app de relatório (atividade, comparacao, etc.) são registradas diretamente no arquivo principal config/urls.py para maior clareza, como você observou.

Organização do settings.py: A lista de INSTALLED_APPS no arquivo config/settings.py foi reorganizada, agrupando as aplicações em categorias (Django, Terceiros e Locais) para melhorar a legibilidade do projeto.

Análises Adicionais:
Particionamento da Tabela: Sobre a sua observação de particionar a tabela, confirmei que a implementação já seguia essa boa prática. O template _tabela_e_cards.html funciona como um partial carregado dinamicamente pelo HTMX. Manter essa separação é o que garante a atualização parcial da página, que é o principal benefício do uso do HTMX aqui. - além de ficar mais simples a implementação futuras para certos filtros que essa página pedir.

caiqueBTC committed:
refactor(relatorios): resolve conflitos seting/url e service

caiqueBTC committed:
ATHOS-15 refactor(relatorios): resolve conflitos seting/url

rtrevizoli commented:
Funcionalidades testadas e validades, estou aguardando a atualização da branch com a develop e a resolução de conflitos para aprovar.

rtrevizoli committed:
ATHOS-15 chore(PR): Atualiza branch com a develop

Merge branch 'develop' into ATHOS-15/Criar-tabela-total-de-horas-consolidadas-por-projeto-front-integracao

rtrevizoli committed:
ATHOS-15 chore(PR): Remove settings adicionado no commit anterior

rtrevizoli committed:
ATHOS-15 chore(relarorio/atividade/services.py): Padronizado params

* Realizado padronização nos parâmetros de entrada dos métodos:
horas_por_dev_e_projeto_por_mes e soma_horas_por_dev_por_mes para
acompanhar formato do filtro.

rtrevizoli reviewed (approved):
LGTM
